### PR TITLE
102-interface-for-streaming-data-from-data-holder

### PR DESCRIPTION
**Pull request by AdrianMontaguSmartDCSIT**
This PR extends the data holder base class with a stream_data method. SQLDataHolder is now able to stream OtelEvents from the database efficiently by using generator expressions, allowing for large datasets to be handled.

Tests have also been added to test functionality.

### CODE REVIEW

File: tel2puml/otel_to_pv/data_holders/sql_data_holder/sql_dataholder.py [hunk 1]
```diff
@@ -230,11 +226,79 @@ def find_unique_graphs(self) -> dict[str, set[str]]:
...
+            # Order by job_name and job_id to use groupby later on.
+            # Limit query object to batch_size
+            query = query.order_by(
+                NodeModel.job_name, NodeModel.job_id
+            ).yield_per(self.batch_size)
+
+            for node in query:
+                job_name = node.job_name
+                otel_event = self.node_to_otel_event(node)
+                yield job_name, otel_event
```
Comment (FreddieMatherSmartDCSIT) [inline tel2puml/otel_to_pv/data_holders/sql_data_holder/sql_dataholder.py:289]: don't need to have job name here as job name is on the otel event anyway and is mandatory

File: tel2puml/otel_to_pv/data_holders/sql_data_holder/sql_dataholder.py [hunk 2]
```diff
@@ -230,11 +226,79 @@ def find_unique_graphs(self) -> dict[str, set[str]]:
...
+
+        with self.session as session:
+            # Get a generator of (job_name, OTelEvent) tuples
+            job_name_event_generator = stream_job_name_batches(
+                session, job_name_to_job_ids_map, filter_job_names
+            )
+            # Group the tuples by job_name
+            grouped = groupby(job_name_event_generator, key=lambda x: x[0])
```
Comment (FreddieMatherSmartDCSIT) [inline tel2puml/otel_to_pv/data_holders/sql_data_holder/sql_dataholder.py:297]: can just use x.job_name instead

File: tests/tel2puml/otel_to_pv/data_holders/sql_data_holder/test_sql_dataholder.py [hunk 3]
```diff
@@ -703,3 +695,158 @@ def test_find_unique_graphs(
...
+    # Test 2: Filter by job_name
+    result = sql_data_holder.stream_data(filter_job_names={"test_name_0"})
+    expected_job_names = ["test_name_0"]
+    all_events = []
+    for job_name, otel_event_gen in result:
+        assert job_name in expected_job_names
+        expected_job_names.remove(job_name)
+
+        for otel_event in otel_event_gen:
+            all_events.append(otel_event)
```
Comment (FreddieMatherSmartDCSIT) [inline tests/tel2puml/otel_to_pv/data_holders/sql_data_holder/test_sql_dataholder.py:738]: think a check to make sure the otel event has the required job name would be good here

File: tests/tel2puml/otel_to_pv/data_holders/sql_data_holder/test_sql_dataholder.py [hunk 4]
```diff
@@ -703,3 +695,158 @@ def test_find_unique_graphs(
...
+            "test_name_1": {"test_id_2", "test_id_3", "test_id_4"},
+        },
+    )
+    job_event_counts = {}
+    all_events = []
+    for job_name, otel_event_gen in result:
+        job_event_counts[job_name] = 0
+        for otel_event in otel_event_gen:
+            all_events.append(otel_event)
+            job_event_counts[job_name] += 1
```
Comment (FreddieMatherSmartDCSIT) [inline tests/tel2puml/otel_to_pv/data_holders/sql_data_holder/test_sql_dataholder.py:756]: again check for correct job name would be good. We also need a check for correct job id somewhere

File: tests/tel2puml/otel_to_pv/data_holders/sql_data_holder/test_sql_dataholder.py [hunk 5]
```diff
@@ -703,3 +695,158 @@ def test_find_unique_graphs(
     assert unique_job_ids_per_job_name["test_name_2"] == {
         str(i) for i in range(5)
     }
+
+
+def test_stream_data(
```
Comment (FreddieMatherSmartDCSIT) [inline tests/tel2puml/otel_to_pv/data_holders/sql_data_holder/test_sql_dataholder.py:700]: I think once you test the sub function to this one that was mentioned in a previous comment, some of the tests can probably done for that function.

**Review by FreddieMatherSmartDCSIT — CHANGES_REQUESTED** [review]
Main thing is to extract the function stream_job_name_batches that is doing a lot of the work from inside stream_data and test the functionality of that. Some of the tests for that will mean you don't have to test those cases in stream_data.

File: tests/tel2puml/otel_to_pv/conftest.py [hunk 6]
```diff
@@ -1020,6 +1020,54 @@ def otel_jobs() -> dict[str, list[OTelEvent]]:
     return otel_jobs
 
 
+@pytest.fixture
+def otel_jobs_multiple_job_names() -> dict[str, list[OTelEvent]]:
```
[the remark this answers is not in the document]
Comment (AdrianMontaguSmartDCSIT) [reply tests/tel2puml/otel_to_pv/conftest.py]: Changes made

File: tel2puml/otel_to_pv/data_holders/sql_data_holder/sql_dataholder.py (hunk 2 above)
Comment (AdrianMontaguSmartDCSIT) [reply tel2puml/otel_to_pv/data_holders/sql_data_holder/sql_dataholder.py:297]: Resolved

File: tests/tel2puml/otel_to_pv/data_holders/sql_data_holder/test_sql_dataholder.py [hunk 7]
```diff
@@ -703,3 +695,168 @@ def test_find_unique_graphs(
...
+        otel_event_gen = sql_data_holder.stream_job_name_batches(
+            session,
+            job_name_to_job_ids_map={"test_name": {"test_id_0", "test_id_1"}},
+        )
+        all_events = list(otel_event_gen)
+
+    assert len(all_events) == 4
+    for otel_event in all_events:
+        assert otel_event.job_name == "test_name"
+        assert otel_event.job_id in {
```
Comment (FreddieMatherSmartDCSIT) [inline tests/tel2puml/otel_to_pv/data_holders/sql_data_holder/test_sql_dataholder.py:821]: how many of each job id are you expecting and also you should know the event ids as well. Just leaves no doubt that what you've produced is what you're expecting

File: tests/tel2puml/otel_to_pv/data_holders/sql_data_holder/test_sql_dataholder.py [hunk 8]
```diff
@@ -703,3 +695,168 @@ def test_find_unique_graphs(
...
+            "test_id_0",
+            "test_id_1",
+        }
+
+    # Test 4: Stream data, filtering with name and map
+    filter_job_names = {"test_name"}
+    job_name_to_job_ids_map = {"test_name": {"test_id_0", "test_id_1"}}
```
Comment (FreddieMatherSmartDCSIT) [inline tests/tel2puml/otel_to_pv/data_holders/sql_data_holder/test_sql_dataholder.py:828]: should have test_name_1 in the dictionary as I can't tell that the filter is working with the map here

File: tel2puml/otel_to_pv/data_holders/sql_data_holder/sql_dataholder.py [hunk 9]
```diff
@@ -230,12 +226,77 @@ def find_unique_graphs(self) -> dict[str, set[str]]:
...
+    ) -> Generator[OTelEvent, None, None]:
+        """
+        Stream OTelEvents from the database.
+
+        :param session: SQLAlchemy Session object.
+        :type session: :class: `sqlalchemy.orm.Session`
+        :param job_name_to_job_ids_map: Optional mapping of job names to
+        job IDs to filter.
+        :type job_name_to_job_ids_map: `dict`[`str`, `set`[`str`]] | `None`
+        :param filter_job_names: Optional set of job names to filter.
```
Comment (FreddieMatherSmartDCSIT) [inline tel2puml/otel_to_pv/data_holders/sql_data_holder/sql_dataholder.py:245]: In future can you put what it defaults as thats how you write sphinx type docs i.e. :param filter_job_names: Optional set of job names to filter, defaults to None
and below :type filter_job_names: `set`[`str`] | `None`, optional
can you change this throughout

**Review by FreddieMatherSmartDCSIT — CHANGES_REQUESTED** [review]
Just a couple more things and then can merge

File: tests/tel2puml/otel_to_pv/data_holders/sql_data_holder/test_sql_dataholder.py [hunk 10]
```diff
@@ -703,3 +695,201 @@ def test_find_unique_graphs(
...
+    sql_data_holder_with_multiple_otel_job_names: SQLDataHolder,
+) -> None:
+    """Test stream_job_name_batches method."""
+    # Test 1: Stream all data
+    sql_data_holder = sql_data_holder_with_multiple_otel_job_names
+    with sql_data_holder.session as session:
+        otel_event_gen = sql_data_holder.stream_job_name_batches(session)
+        all_events = list(otel_event_gen)
+
+    assert len(all_events) == 20
```
Comment (FreddieMatherSmartDCSIT) [inline tests/tel2puml/otel_to_pv/data_holders/sql_data_holder/test_sql_dataholder.py:801]: just for this one can we make sure we have the sequence of job names and job ids correct as we should be checking that its all correctly ordered in the output i.e.

```
assert all(
    event.job_name == expected_job_name and event.job_id == expected_job_id
    for event, expected_job_name, expected_job_id in zip(
        all_events,
        ["test_name"] * 10 + ["test_name_1"] *10),
        [f"test_id_{i}" for i in range(10) for _ in range(2)]
)
```

**Review by FreddieMatherSmartDCSIT — CHANGES_REQUESTED** [review]
Just test of the sequence of job names and job ids to do - as this is what is required for the other function

File: tests/tel2puml/otel_to_pv/data_holders/sql_data_holder/test_sql_dataholder.py (hunk 10 above)
Comment (AdrianMontaguSmartDCSIT) [reply tests/tel2puml/otel_to_pv/data_holders/sql_data_holder/test_sql_dataholder.py:801]: Change made

**Review by FreddieMatherSmartDCSIT — APPROVED** [review]
Looks good to me!

**Pull request merged**